library/common/http/client.cc

```diff
@@ -117,19 +164,63 @@ void Client::DirectStreamCallbacks::encodeTrailers(const ResponseTrailerMap& tra
   ASSERT(http_client_.getStream(direct_stream_.stream_handle_));
   closeStream(); // Trailers always indicate the end of the stream.
 
+  // When explicitly buffering, don't send data unless prompted.
+  if (explicit_buffering_ && bytes_to_send_ == 0) {
+    response_trailers_ = ResponseTrailerMapImpl::create();
+    HeaderMapImpl::copyFrom(*response_trailers_, trailers);
+    return;
+  }
+
+  sendTrailersToBridge(trailers);
+}
+
+void Client::DirectStreamCallbacks::sendTrailersToBridge(const ResponseTrailerMap& trailers) {
   ENVOY_LOG(debug, "[S{}] dispatching to platform response trailers for stream:\n{}",
             direct_stream_.stream_handle_, trailers);
   bridge_callbacks_.on_trailers(Utility::toBridgeHeaders(trailers), bridge_callbacks_.context);
   onComplete();
 }
 
+void Client::DirectStreamCallbacks::resumeData(int32_t bytes_to_send) {
+  ASSERT(explicit_buffering_);
+  ASSERT(bytes_to_send > 0);
+
+  bytes_to_send_ = bytes_to_send;
+
+  ENVOY_LOG(debug, "[S{}] received resume data call for {} bytes", direct_stream_.stream_handle_,
+            bytes_to_send_);
+
+  // If there is bufferd data, send up to bytes_to_send bytes.
+  // Make sure to send end stream with data only if
+  // 1) it has been received from the peer and
+  // 2) there are no trailers
+  if (hasBufferedData() || (end_stream_read_ && !end_stream_communicated_ && !response_trailers_)) {
```

Comment:
Should some or all of this maybe just be an ASSERT? For example, we should never have received end_stream on a data frame if we received trailers. Similarly, I saw that you're guarding elsewhere that you don't forward end_stream if you need to buffer data.

Reply:
I think these are legit conditions :-)
we get "end stream" either with a data FIN or with trailers so this is basically saying "if we got end stream not via trailers" send up empty-data-with-fin.
For sendDataToBridge if we have 12k+fin buffered (no trailers) and we're only allowed to send 8k up we need to not send the end stream ahead of the last 4k

Reply:
Got it - I initially misread this as more of a guard than valid branch.